### PR TITLE
KAFKA-16181: Use incrementalAlterConfigs when updating broker configs by kafka-configs.sh

### PR DESCRIPTION
**Pull request by dengziming**
*More detailed description of your change*
This PR implement KIP-1011, use incrementalAlterConfigs by default to alter broker configurations instead of the deprecated alterConfigs API, and fall back to deprecated alterConfigs API if the broker doesn't support incrementalAlterConfigs API.

*Summary of testing strategy (including rationale)*
1. Add some test cases to verify the new approach.
2. TODO, we can't verify the automatic process because we can't produce `UnsupportedVersionException` by setting inter.broker.protocol.version

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 1]
```diff
@@ -368,14 +368,29 @@ object ConfigCommand extends Logging {
...
         val alterOptions = new AlterConfigsOptions().timeoutMs(30000).validateOnly(false)
         val alterEntries = (configsToBeAdded.values.map(new AlterConfigOp(_, AlterConfigOp.OpType.SET))
           ++ configsToBeDeleted.map { k => new AlterConfigOp(new ConfigEntry(k, ""), AlterConfigOp.OpType.DELETE) }
         ).asJavaCollection
-        adminClient.incrementalAlterConfigs(Map(configResource -> alterEntries).asJava, alterOptions).all().get(60, TimeUnit.SECONDS)
 
-      case ConfigType.BROKER =>
+        var retryUsingDeprecatedAlterConfigs = false
+        adminClient.incrementalAlterConfigs(Map(configResource -> alterEntries).asJava, alterOptions).all().whenComplete((_, e) => {
+          if (e.isInstanceOf[UnsupportedVersionException] && entityTypeHead == ConfigType.BROKER) {
```
Comment (OmniaGM) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala]: Usually in scala `isInstanceOf` is kinda of anti-pattern and mostly used in tests but not in prod code. Using pattern matching may be better option here; something like the following
 
```
        .whenComplete {
            case (_ , _: UnsupportedVersionException) if entityTypeHead == ConfigType.BROKER =>
              retryUsingDeprecatedAlterConfigs = true
            case _ => // nothing to do
          }
   ```

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 2]
```diff
@@ -405,7 +420,7 @@ object ConfigCommand extends Logging {
         val alterOptions = new AlterConfigsOptions().timeoutMs(30000).validateOnly(false)
         val alterLogLevelEntries = (configsToBeAdded.values.map(new AlterConfigOp(_, AlterConfigOp.OpType.SET))
           ++ configsToBeDeleted.map { k => new AlterConfigOp(new ConfigEntry(k, ""), AlterConfigOp.OpType.DELETE) }
-        ).asJavaCollection
+          ).asJavaCollection
```
Comment (OmniaGM) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala]: this change maybe not needed!

File: core/src/test/scala/integration/kafka/admin/ConfigCommandIntegrationTest.scala [hunk 3]
```diff
@@ -162,19 +174,198 @@ class ConfigCommandIntegrationTest extends QuorumTestHarness with Logging {
...
+  def testUpdateInvalidBrokersConfig(): Unit = {
+    checkInvalidBrokerConfig(None)
+    checkInvalidBrokerConfig(Some(cluster.anyBrokerSocketServer().config.brokerId.toString))
+  }
+
+  private def checkInvalidBrokerConfig(entityNameOrDefault: Option[String]): Unit = {
+    for (incremental <- Array(true, false)) {
+      val entityNameParams = entityNameOrDefault.map(name => Array("--entity-name", name)).getOrElse(Array("--entity-default"))
+      ConfigCommand.alterConfig(cluster.createAdminClient(), new ConfigCommandOptions(
+        Array("--bootstrap-server", s"${cluster.bootstrapServers()}",
```
Comment (OmniaGM) [inline core/src/test/scala/integration/kafka/admin/ConfigCommandIntegrationTest.scala]: There is no need to use string interpolation here as `cluster.bootstrapServers()` is already a string.

File: core/src/test/scala/integration/kafka/admin/ConfigCommandIntegrationTest.scala [hunk 4]
```diff
@@ -162,19 +174,198 @@ class ConfigCommandIntegrationTest extends QuorumTestHarness with Logging {
...
+          "--alter",
+          "--add-config", "invalid=2",
+          "--entity-type", "brokers")
+          ++ entityNameParams
+      ), incremental)
+
+      val describeResult = TestUtils.grabConsoleOutput(
+        ConfigCommand.describeConfig(cluster.createAdminClient(), new ConfigCommandOptions(
+          Array("--bootstrap-server", s"${cluster.bootstrapServers()}",
```
Comment (OmniaGM) [inline core/src/test/scala/integration/kafka/admin/ConfigCommandIntegrationTest.scala]: Same need to be updated for all test cases

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 5]
```diff
@@ -349,7 +349,7 @@ object ConfigCommand extends Logging {
   }
 
   @nowarn("cat=deprecation")
-  private[admin] def alterConfig(adminClient: Admin, opts: ConfigCommandOptions): Unit = {
+  private[admin] def alterConfig(adminClient: Admin, opts: ConfigCommandOptions, useIncrementalAlterConfigs: Boolean): Unit = {
```
Comment (OmniaGM) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala]: couldn't we set `useIncrementalAlterConfigs` to `true` as default specially as I can see that later we set it with true any way.

File: core/src/test/scala/unit/kafka/admin/ConfigCommandTest.scala [hunk 6]
```diff
@@ -430,7 +430,7 @@ class ConfigCommandTest extends Logging {
   def shouldFailIfUnrecognisedEntityType(): Unit = {
     val createOpts = new ConfigCommandOptions(Array("--bootstrap-server", "localhost:9092",
       "--entity-name", "client", "--entity-type", "not-recognised", "--alter", "--add-config", "a=b,c=d"))
-    assertThrows(classOf[IllegalArgumentException], () => ConfigCommand.alterConfig(new DummyAdminClient(new Node(1, "localhost", 9092)), createOpts))
+    assertThrows(classOf[IllegalArgumentException], () => ConfigCommand.alterConfig(new DummyAdminClient(new Node(1, "localhost", 9092)), createOpts, true))
```
Comment (OmniaGM) [inline core/src/test/scala/unit/kafka/admin/ConfigCommandTest.scala]: if `useIncrementalAlterConfigs` set to true as default we wouldn't need to explicitly set it to true.

File: core/src/test/scala/unit/kafka/admin/ConfigCommandTest.scala [hunk 7]
```diff
@@ -1684,7 +1684,7 @@ class ConfigCommandTest extends Logging {
         alterResult
       }
     }
-    ConfigCommand.alterConfig(mockAdminClient, alterOpts)
+    ConfigCommand.alterConfig(mockAdminClient, alterOpts, true)
     verify(describeResult).all()
     verify(alterResult).all()
   }
```
Comment (OmniaGM) [inline core/src/test/scala/unit/kafka/admin/ConfigCommandTest.scala]: I can't see any test for `alterConfig` with `useIncrementalAlterConfigs` set to false. Can we add some tests for this path as well if it is possible?

File: core/src/test/scala/integration/kafka/admin/ConfigCommandIntegrationTest.scala [hunk 8]
```diff
@@ -162,19 +174,198 @@ class ConfigCommandIntegrationTest extends QuorumTestHarness with Logging {
...
+  @ClusterTest(clusterType=Type.ZK, metadataVersion = MetadataVersion.IBP_3_2_IV0)
+  def testFallbackToDeprecatedAlterConfigs(): Unit = {
+    ConfigCommand.alterConfig(cluster.createAdminClient(), new ConfigCommandOptions(
+      Array("--bootstrap-server", s"${cluster.bootstrapServers()}",
+        "--alter",
+        "--add-config", "log.cleaner.threads=2",
+        "--entity-type", "brokers",
+        "--entity-default")
+    ), true)
+  }
```
Comment (OmniaGM) [inline core/src/test/scala/integration/kafka/admin/ConfigCommandIntegrationTest.scala]: I can't see any tests for `useIncrementalAlterConfigs` set to false. Can we test this path as well if possible?

**Review by OmniaGM — COMMENTED** [review]
Thanks for the PR. I left few small comments

File: core/src/main/scala/kafka/admin/ConfigCommand.scala (hunk 5 above)
Comment (dengziming) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala]: Good catch, I added a case of updating broker configs with `useIncrementalAlterConfigs=false`, also add a case of updating topic configs which only support `incrementalAlterConfigs`.

File: core/src/test/scala/integration/kafka/admin/ConfigCommandIntegrationTest.scala (hunk 8 above)
Comment (dengziming) [reply core/src/test/scala/integration/kafka/admin/ConfigCommandIntegrationTest.scala]: This case doesn't make sense because we can't produce `UnsupportedVersionException` by setting inter.broker.protocol.version, so I'm waiting for help from other reviewers, and just leave it unchanged for now.

Comment (dajac) [reply core/src/test/scala/integration/kafka/admin/ConfigCommandIntegrationTest.scala]: @dengziming One option would be to mock the admin client. This would allow you to verify the behavior. Otherwise, we could also run the tool against an old version in a system test.

**Comment by dengziming** [comment]
@dajac Hello, this PR is ready for review except the upgrade docs, PTAL.

**Comment by dengziming** [comment]
Hi @dajac , this PR has been blocked for a long time, and the test case has been rewritten using java, I rebased my code and changed the test code to java too, PTAL 🤝.

**Comment by frankvicky** [comment]
Hi @dengziming 
Could you please fix the fail in CI ?

**Comment by dengziming** [comment]
@frankvicky Thank you, the fail is related to some small test bugs, I will fix them in #17407

File: core/src/main/scala/kafka/admin/ConfigCommand.scala (hunk 2 above)
Comment (dengziming) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala]: This is done automatically by the scalastyle check, I think this is reasonable according to the style suggestions.

**Comment by dengziming** [comment]
Hi @dajac @showuon , as we discussed in the email thread, I directly removed the deprecated `alterConfigs` in `ConfigCommand`, it's very concise of the changes now, PTAL, and I will add this to 4.0 release plan and alter the KIP-1011 pages

File: core/src/main/scala/kafka/admin/ConfigCommand.scala (hunk 2 above)
Comment (dajac) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala]: I find this line pretty hard to read. While we do changes here, I wonder if we could refactor it a bit more. What do you think?

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 9]
```diff
@@ -290,17 +293,87 @@ public void testUpdatePerBrokerConfigInKRaftThenShouldFail() {
...
+    }
+
+    // Test case from KAFKA-13788
+    @ClusterTest
+    public void testUpdateBrokerConfigNotAffectedByInvalidConfig() {
+        try (Admin client = cluster.admin()) {
+            ConfigCommand.alterConfig(client, new ConfigCommand.ConfigCommandOptions(
+                    toArray(asList("--bootstrap-server", cluster.bootstrapServers(),
+                            "--alter",
+                            "--add-config", "log.cleaner.threadzz=2",
```
Comment (dajac) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:369]: For my understanding, it seems that we don't validate broker configs. I suppose that we don't because plugins could rely on unknown configs. Is my understanding correct?

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 10]
```diff
@@ -290,17 +293,87 @@ public void testUpdatePerBrokerConfigInKRaftThenShouldFail() {
...
+            assertTrue(describeResult.contains("invalid=null"), describeResult);
+        }
+    }
+
+    @ClusterTest
+    public void testUpdateInvalidTopicConfigs() throws ExecutionException, InterruptedException {
+        List<String> alterOpts = asList("--bootstrap-server", cluster.bootstrapServers(), "--entity-type", "topics", "--alter");
+        try (Admin client = cluster.admin()) {
+            client.createTopics(Collections.singletonList(new NewTopic("test-config-topic", 1, (short) 1))).all().get();
+            assertInstanceOf(
```
Comment (dajac) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:345]: nit: You may be able to use `assertFutureThrows`.

File: core/src/test/java/kafka/admin/ConfigCommandTest.java [hunk 11]
```diff
@@ -1019,13 +1019,13 @@ public synchronized DescribeConfigsResult describeConfigs(Collection<ConfigResou
 
             @SuppressWarnings("deprecation")
```
Comment (dajac) [inline core/src/test/java/kafka/admin/ConfigCommandTest.java]: I suppose that it does not apply any more.

File: core/src/test/scala/unit/kafka/utils/TestUtils.scala [hunk 12]
```diff
@@ -333,6 +333,9 @@ object TestUtils extends Logging {
     if (!props.containsKey(GroupCoordinatorConfig.GROUP_INITIAL_REBALANCE_DELAY_MS_CONFIG))
       props.put(GroupCoordinatorConfig.GROUP_INITIAL_REBALANCE_DELAY_MS_CONFIG, "0")
     rack.foreach(props.put(ServerConfigs.BROKER_RACK_CONFIG, _))
+    // reduce log cleaner offset map memory usage, must be at greater than 1MB per cleaner thread, set to 2M+2 so that
+    // we can set 2 cleaner threads.
+    props.put(CleanerConfig.LOG_CLEANER_DEDUPE_BUFFER_SIZE_PROP, "2097154")
```
Comment (dajac) [inline core/src/test/scala/unit/kafka/utils/TestUtils.scala]: Is this change related to the PR?

File: test-common/src/main/java/org/apache/kafka/common/test/KafkaClusterTestKit.java [hunk 13]
```diff
@@ -174,8 +174,9 @@ private KafkaConfig createNodeConfig(TestKitNode node) throws IOException {
             }
             props.put(QuorumConfig.QUORUM_VOTERS_CONFIG, quorumVoterStringBuilder.toString());
 
-            // reduce log cleaner offset map memory usage
-            props.put(CleanerConfig.LOG_CLEANER_DEDUPE_BUFFER_SIZE_PROP, "2097152");
+            // reduce log cleaner offset map memory usage, must be at greater than 1MB per cleaner thread, set to 2M+2 so that
+            // we can set 2 cleaner threads.
+            props.put(CleanerConfig.LOG_CLEANER_DEDUPE_BUFFER_SIZE_PROP, "2097154");
```
Comment (dajac) [inline test-common/src/main/java/org/apache/kafka/common/test/KafkaClusterTestKit.java]: Is this change related to the PR?

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 14]
```diff
@@ -172,27 +171,14 @@ object ConfigCommand extends Logging {
...
-        val alterOptions = new AlterConfigsOptions().timeoutMs(30000).validateOnly(false)
-        adminClient.alterConfigs(Map(configResource -> newConfig).asJava, alterOptions).all().get(60, TimeUnit.SECONDS)
+      case ConfigType.TOPIC | ConfigType.CLIENT_METRICS | ConfigType.BROKER | ConfigType.GROUP =>
+        val configResourceType = entityTypeHead match {
+          case ConfigType.TOPIC => ConfigResource.Type.TOPIC
+          case ConfigType.CLIENT_METRICS =>ConfigResource.Type.CLIENT_METRICS
+          case ConfigType.BROKER => ConfigResource.Type.BROKER
+          case ConfigType.GROUP => ConfigResource.Type.GROUP
+        }
+        alterResourceConfig(adminClient, entityTypeHead, entityNameHead, configsToBeDeleted, configsToBeAdded, configResourceType)
```
Comment (dajac) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala]: What does happen if the unsupported exception is thrown here? I wonder whether we should catch it in the main method in order to provide a meaningful error message to the user. What do you think?

**Review by dajac — COMMENTED** [review]
@dengziming Thanks for the patch! I left a few comments for consideration.

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java (hunk 9 above)
Comment (dengziming) [reply core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:369]: Currently we treat unknown broker configs as sensitive configs and do not validate them, not sure whether it's related to plugins.

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java (hunk 10 above)
Comment (dengziming) [reply core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java:345]: I tried but the parameter type is not consistent, `assertFutureThrows ` is mostly used to check result from a `AdminClient` not a `Command.main`, and I find there are 2 assertFutureThrows in TestUtils which can be merged in the future.

File: core/src/test/scala/unit/kafka/utils/TestUtils.scala (hunk 12 above)
Comment (dengziming) [reply core/src/test/scala/unit/kafka/utils/TestUtils.scala]: Yes, I update this config from 1 to 2 in my test cases, which would fail without this change.

File: test-common/src/main/java/org/apache/kafka/common/test/KafkaClusterTestKit.java (hunk 13 above)
Comment (dengziming) [reply test-common/src/main/java/org/apache/kafka/common/test/KafkaClusterTestKit.java]: ditto

File: core/src/main/scala/kafka/admin/ConfigCommand.scala (hunk 14 above)
Comment (dengziming) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala]: That's a nice catch, I added exception handling logic and a test case, I also started a 2.0 kafka cluster to validate it, I pasted the screenshot below.

File: core/src/main/scala/kafka/admin/ConfigCommand.scala (hunk 2 above)
Comment (dengziming) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala]: This fragment also duplicate in `alterResourceConfig`, I changed them all to be clear.

**Review by dengziming — COMMENTED** [review]
img_v3_02h1_5ae695b8-1267-4cff-9ff1-0c7a8c7c5c3g

Thank you @dajac for your review, I also validated the case locally in a 2.0 Kafka cluster, PTAL again.

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 15]
```diff
@@ -172,27 +173,26 @@ object ConfigCommand extends Logging {
...
+        }
+        try {
+          alterResourceConfig(adminClient, entityTypeHead, entityNameHead, configsToBeDeleted, configsToBeAdded, configResourceType)
+        } catch {
+          case e: ExecutionException =>
+            e.getCause match {
+              case _: UnsupportedVersionException if entityTypeHead == ConfigType.BROKER =>
+                System.err.println(s"Could not update broker config $entityNameHead, because brokers don't support api ${ApiKeys.INCREMENTAL_ALTER_CONFIGS},"
+                + " You can upgrade your brokers to version 2.3.0 or newer to avoid this error.")
+                return
```
Comment (dajac) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala]: I find it a little weird that we log the error here vs doing it in the `main ` function. Is there a particular reason for doing it this way? My concern is that it does not exit as it should with `Exit.exit(1)`. Could we catch it there by adding a new case? An alternative would be to still catch it here but to rethrow an `UnsupportedVersionException` with the message that we want and catch it in the main. What do you think?

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 16]
```diff
@@ -172,27 +173,26 @@ object ConfigCommand extends Logging {
...
-          throw new InvalidConfigurationException(s"All sensitive broker config entries must be specified for --alter, missing entries: ${sensitiveEntries.keySet}")
-        val newConfig = new JConfig(newEntries.asJava.values)
-
-        val configResource = new ConfigResource(ConfigResource.Type.BROKER, entityNameHead)
-        val alterOptions = new AlterConfigsOptions().timeoutMs(30000).validateOnly(false)
-        adminClient.alterConfigs(Map(configResource -> newConfig).asJava, alterOptions).all().get(60, TimeUnit.SECONDS)
+      case ConfigType.TOPIC | ConfigType.CLIENT_METRICS | ConfigType.BROKER | ConfigType.GROUP =>
+        val configResourceType = entityTypeHead match {
+          case ConfigType.TOPIC => ConfigResource.Type.TOPIC
+          case ConfigType.CLIENT_METRICS =>ConfigResource.Type.CLIENT_METRICS
```
Comment (dajac) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala]: nit: A space misses after `=>`.

File: core/src/test/scala/unit/kafka/utils/TestUtils.scala (hunk 12 above)
Comment (dajac) [reply core/src/test/scala/unit/kafka/utils/TestUtils.scala]: Is this something that we could set at the test level? You can pass extra configs to `@ClusterTest`.

**Review by dajac — COMMENTED** [review]
@dengziming Thanks for the update. I left a few more comments.

File: core/src/test/scala/unit/kafka/utils/TestUtils.scala (hunk 12 above)
Comment (dengziming) [reply core/src/test/scala/unit/kafka/utils/TestUtils.scala]: This approach is more reasonable, I reverted the changes in TestUtils and KafkaClusterTestKit, but we need to change `props.put` to `props.putIfAbsent` in  KafkaClusterTestKit since it will overwrite the original value.
Outdated

File: core/src/main/scala/kafka/admin/ConfigCommand.scala (hunk 15 above)
Comment (dengziming) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala]: Catching it here is more accurate just in case of other unsupported-error from the server, and it should be handled in the main function, I wrapped a new Exception and rethrow to main causing the process to exit with code 1, this seems more reasonalbe.
image

**Review by dengziming — COMMENTED** [review]
The 2 suggestions are nice, I made some improvements, plz take a look again. @dajac

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 17]
```diff
@@ -80,6 +82,11 @@ object ConfigCommand extends Logging {
         System.err.println(e.getMessage)
         Exit.exit(1)
 
+      case e: UnsupportedVersionException =>
+        logger.debug(s"Unsupported API encountered in server when executing config command with args '${args.mkString(" ")}'")
+        e.printStackTrace(System.err)
```
Comment (dajac) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala]: nit: It may be better to use `System.err.println(e.getMessage)`. What do you think?

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 18]
```diff
@@ -172,27 +178,25 @@ object ConfigCommand extends Logging {
...
+          case ConfigType.CLIENT_METRICS => ConfigResource.Type.CLIENT_METRICS
+          case ConfigType.BROKER => ConfigResource.Type.BROKER
+          case ConfigType.GROUP => ConfigResource.Type.GROUP
+        }
+        try {
+          alterResourceConfig(adminClient, entityTypeHead, entityNameHead, configsToBeDeleted, configsToBeAdded, configResourceType)
+        } catch {
+          case e: ExecutionException =>
+            e.getCause match {
+              case cause: UnsupportedVersionException if entityTypeHead == ConfigType.BROKER =>
```
Comment (dajac) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala]: I wonder if we could remove the `entityTypeHead == ConfigType.BROKER` condition to cover all the types. Is there a reason not to?

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 19]
```diff
@@ -172,27 +178,25 @@ object ConfigCommand extends Logging {
...
+                throw new UnsupportedVersionException(s"Could not update broker config $entityNameHead, because brokers don't support api ${ApiKeys.INCREMENTAL_ALTER_CONFIGS},"
+                + " You can upgrade your brokers to version 2.3.0 or newer to avoid this error.", cause)
```
Comment (dajac) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala]: nit: I would say something like this: `The ${ApiKeys.INCREMENTAL_ALTER_CONFIGS} API is not supported by the cluster. The API is supported starting from version 2.3.0. You may want to use an older version of this tool to interact with your cluster.`. It seems better to me to suggest to use an older tool because I suppose that the user won't jump on upgrading his cluster.

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 20]
```diff
@@ -290,17 +303,125 @@ public void testUpdatePerBrokerConfigInKRaftThenShouldFail() {
...
+                    100L);
+        }
+    }
+
+    @ClusterTest(
+            serverProperties = {@ClusterConfigProperty(key = "log.cleaner.dedupe.buffer.size", value = "2097154")},
+            // Zk code has been removed, use kraft and mockito to mock this situation
+            metadataVersion = MetadataVersion.IBP_3_3_IV0
+    )
+    public void testFallbackToDeprecatedAlterConfigs() throws NoSuchMethodException, InvocationTargetException, InstantiationException, IllegalAccessException {
```
Comment (dajac) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: nit: The name is misleading because it suggests that we have a fallback mechanism.

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 21]
```diff
@@ -290,17 +303,125 @@ public void testUpdatePerBrokerConfigInKRaftThenShouldFail() {
...
+                                            toArray(alterOpts,
+                                                    asList("--add-config", "invalid=2", "--entity-type", "topics", "--entity-name", "test-config-topic"))))
+                    ).getCause()
+            );
+        }
+    }
+
+    // Test case from KAFKA-13788
+    @ClusterTest(serverProperties = {
+            @ClusterConfigProperty(key = "log.cleaner.dedupe.buffer.size", value = "2097154"),
```
Comment (dajac) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: nit: We usually indent with 4 spaces.

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 22]
```diff
@@ -290,17 +303,125 @@ public void testUpdatePerBrokerConfigInKRaftThenShouldFail() {
...
+                    () -> "Timeout waiting for topic config propagating to broker",
+                    org.apache.kafka.test.TestUtils.DEFAULT_MAX_WAIT_MS,
+                    100L);
+        }
+    }
+
+    @ClusterTest(
+            serverProperties = {@ClusterConfigProperty(key = "log.cleaner.dedupe.buffer.size", value = "2097154")},
+            // Zk code has been removed, use kraft and mockito to mock this situation
+            metadataVersion = MetadataVersion.IBP_3_3_IV0
```
Comment (dajac) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: ditto.

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java (hunk 21 above)
Comment (dajac) [reply core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: nit: Could also put a comment explaining why we need to change this config. It is not obvious.

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java [hunk 23]
```diff
@@ -290,17 +303,125 @@ public void testUpdatePerBrokerConfigInKRaftThenShouldFail() {
...
+            AlterConfigsResult mockResult;
+            {
+                // Create a mock result of unsupported version error
+                KafkaFutureImpl<Void> future = new KafkaFutureImpl<>();
+                future.completeExceptionally(new UnsupportedVersionException("simulated error"));
+                Constructor<AlterConfigsResult> constructor = AlterConfigsResult.class.getDeclaredConstructor(java.util.Map.class);
+                constructor.setAccessible(true);
+                mockResult = constructor.newInstance(Collections.singletonMap(new ConfigResource(ConfigResource.Type.BROKER, ""), future));
+                constructor.setAccessible(false);
+            }
```
Comment (dajac) [inline core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: nit: You may be able to use AdminClientTestUtils.alterConfigsResult.

**Review by dajac — COMMENTED** [review]
@dengziming Thanks for your patience on this one. I left a few more suggestions.

File: core/src/main/scala/kafka/admin/ConfigCommand.scala (hunk 17 above)
Comment (dengziming) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala]: Yes, since it's an expected error, it's unnecessary to print the full stack trace.

File: core/src/main/scala/kafka/admin/ConfigCommand.scala (hunk 18 above)
Comment (dengziming) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala]: I was trying to fallback to old API so we should only support BROKER, now it's ok to cover all types.

File: core/src/main/scala/kafka/admin/ConfigCommand.scala (hunk 19 above)
Comment (dengziming) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala]: I reworded it to firstly suggest using older version of client, and I still keep the suggestion of upgrading server.

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java (hunk 21 above)
Comment (dengziming) [reply core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: A comment has been added to explain this config, and optimized the indentation.(We don't enforce it in checkstyle.xml and there is not a unified config for us, I find the default value in IDEA is 8, maybe we can set it in the future but that would change too much code).

image

File: core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java (hunk 23 above)
Comment (dengziming) [reply core/src/test/java/kafka/admin/ConfigCommandIntegrationTest.java]: Good suggestions.

**Review by dengziming — COMMENTED** [review]
@dajac Thank you for your patience and preciseness, left some comments, please take a look.

**Review by dajac — APPROVED** [review]
LGTM, thanks @dengziming!

**Comment by dengziming** [comment]
Thanks for the patience @dajac , merge to trunk.

File: core/src/main/scala/kafka/admin/ConfigCommand.scala [hunk 24]
```diff
@@ -172,27 +178,25 @@ object ConfigCommand extends Logging {
...
+          case ConfigType.BROKER => ConfigResource.Type.BROKER
+          case ConfigType.GROUP => ConfigResource.Type.GROUP
+        }
+        try {
+          alterResourceConfig(adminClient, entityTypeHead, entityNameHead, configsToBeDeleted, configsToBeAdded, configResourceType)
+        } catch {
+          case e: ExecutionException =>
+            e.getCause match {
+              case _: UnsupportedVersionException =>
+                throw new UnsupportedVersionException(s"The ${ApiKeys.INCREMENTAL_ALTER_CONFIGS} API is not supported by the cluster. The API is supported starting from version 2.3.0."
```
Comment (chia7712) [inline core/src/main/scala/kafka/admin/ConfigCommand.scala:194]: Does this implementation align with [KIP-1011](https://cwiki.apache.org/confluence/display/KAFKA/KIP-1011%3A+Use+incrementalAlterConfigs+when+updating+broker+configs+by+kafka-configs.sh)? Here’s a relevant quote from KIP-1011 for reference.

> When updating broker config,  instead of using Admin.alterConfigs, we will use Admin.incrementalAlterConfigs and fallback to use alterConfigs automatically if incrementalAlterConfigs is not supported

Please let me know if I’ve missed any important discussions in this PR.

Comment (dengziming) [reply core/src/main/scala/kafka/admin/ConfigCommand.scala:194]: Hi @chia7712 , though this KIP is intended to be landed in release-3.8, it was delayed to 4.0, so whether to maintain this compatible deprecation period has became a question.  
And in the KIP we have declared that we will fallback to use `alterConfigs` and this fallback logic will be removed in 4.0, so we just throw an `UnsupportedVersionException` since we are already planning for 4.0. 

Here is the discussion thread and I will add it to the KIP docs today, sorry for the delay. https://lists.apache.org/thread/m9pvhgfwtr05rf9sdhn22x11cvtbgv6w
